File: backend/app/schemas/notes.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class NoteCreate(BaseModel):
    verse_reference: str = Field(max_length=60)
    note_text: str = Field(max_length=2000)
    verse_id: Optional[uuid.UUID] = None
    translation_id: Optional[uuid.UUID] = None


class NoteUpdate(BaseModel):
    note_text: str = Field(max_length=2000)


class NoteOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    verse_reference: str
    verse_id: Optional[uuid.UUID] = None
    note_text: str
    created_at: datetime
    updated_at: datetime


class HighlightCreate(BaseModel):
    verse_reference: str = Field(max_length=60)
    color: str = Field(pattern="^(yellow|blue|green|red|purple)$")
    verse_id: Optional[uuid.UUID] = None
    translation_id: Optional[uuid.UUID] = None
    verse_start: Optional[int] = None
    verse_end: Optional[int] = None


class HighlightOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    verse_reference: str
    color: str
    verse_id: Optional[uuid.UUID] = None
    verse_start: Optional[int] = None
    verse_end: Optional[int] = None
    created_at: datetime

File: frontend/src/lib/notes.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { api } from '@/lib/api'

export interface Note {
  id: string
  verse_reference: string
  verse_id: string | null
  note_text: string
  created_at: string
  updated_at: string
}

export interface Highlight {
  id: string
  verse_reference: string
  color: 'yellow' | 'blue' | 'green' | 'red' | 'purple'
  verse_id: string | null
  verse_start: number | null
  verse_end: number | null
  created_at: string
}

export function useNotes() {
  return useQuery({ queryKey: ['notes'], queryFn: async () => (await api.get<Note[]>('/notes')).data })
}

export function useHighlights() {
  return useQuery({ queryKey: ['highlights'], queryFn: async () => (await api.get<Highlight[]>('/notes/highlights')).data })
}

export function useCreateNote() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: { verse_reference: string; note_text: string; verse_id?: string; translation_id?: string }) =>
      (await api.post<Note>('/notes', payload)).data,
    onSuccess: () => qc.invalidateQueries({ queryKey: ['notes'] }),
  })
}

export function useUpdateNote() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, note_text }: { id: string; note_text: string }) =>
      (await api.patch<Note>(`/notes/${id}`, { note_text })).data,
    onSuccess: () => qc.invalidateQueries({ queryKey: ['notes'] }),
  })
}

export function useCreateHighlight() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: {
      verse_reference: string
      color: Highlight['color']
      verse_id?: string
      translation_id?: string
      verse_start?: number
      verse_end?: number
    }) => (await api.post<Highlight>('/notes/highlights', payload)).data,
    onSuccess: () => qc.invalidateQueries({ queryKey: ['highlights'] }),
  })
}

export function useDeleteNote() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => api.delete(`/notes/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['notes'] }),
  })
}

export function useDeleteHighlight() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => api.delete(`/notes/highlights/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['highlights'] }),
  })
}
